crates/kilt-dip-primitives/src/merkle/v0/provider_state/mod.rs

```diff
@@ -0,0 +1,495 @@
+	> where
+	RelayBlockNumber: Default,
+	KiltDidKeyId: Default + Clone,
+	KiltAccountId: Clone,
+	KiltBlockNumber: Default + Clone,
+	KiltWeb3Name: Clone,
+	KiltLinkableAccountId: Clone,
+	ConsumerBlockNumber: Default,
+	Context: Clone,
+{
+	fn worst_case(context: Context) -> Self {
+		Self {
+			provider_head_proof: ProviderHeadStateProof::worst_case(context.clone()),
+			dip_commitment_proof: DipCommitmentStateProof::worst_case(context.clone()),
+			dip_proof: DidMerkleProof::worst_case(context.clone()),
+			signature: TimeBoundDidSignature::worst_case(context),
+		}
+	}
+}
+
+impl<
+		RelayBlockNumber,
+		KiltDidKeyId,
+		KiltAccountId,
+		KiltBlockNumber,
+		KiltWeb3Name,
+		KiltLinkableAccountId,
+		ConsumerBlockNumber,
+	>
+	ParachainDipDidProof<
+		RelayBlockNumber,
+		KiltDidKeyId,
+		KiltAccountId,
+		KiltBlockNumber,
+		KiltWeb3Name,
+		KiltLinkableAccountId,
+		ConsumerBlockNumber,
+	> where
+	KiltBlockNumber: BenchmarkDefault,
+{
+	/// Verifies the head data of the state proof for the provider with the
+	/// given para ID and relaychain state root.
+	///
+	/// The generic types indicate the following:
+	/// * `RelayHasher`: The head data hashing algorithm used by the relaychain.
+	/// * `ProviderHeader`: The type of the parachain header to be revealed in
+	///   the state proof.
+	#[allow(clippy::type_complexity)]
+	pub fn verify_provider_head_proof_with_state_root<RelayHasher, ProviderHeader>(
+		self,
+		provider_para_id: u32,
+		relay_state_root: &OutputOf<RelayHasher>,
+	) -> Result<
+		DipDidProofWithVerifiedStateRoot<
+			OutputOf<RelayHasher>,
+			KiltDidKeyId,
+			KiltAccountId,
+			KiltBlockNumber,
+			KiltWeb3Name,
+			KiltLinkableAccountId,
+			ConsumerBlockNumber,
+		>,
+		Error,
+	>
+	where
+		RelayHasher: Hash,
+		ProviderHeader: Decode + HeaderT<Hash = OutputOf<RelayHasher>, Number = KiltBlockNumber>,
+	{
+		let provider_head_storage_key = calculate_parachain_head_storage_key(provider_para_id);
+		// TODO: Figure out why RPC call returns 2 bytes in front which we don't need
+		//This could be the reason (and the solution): https://substrate.stackexchange.com/a/1891/1795
+		let provider_header_result = verify_storage_value_proof_with_decoder::<_, RelayHasher, ProviderHeader>(
+			&provider_head_storage_key,
+			*relay_state_root,
+			self.provider_head_proof.proof,
+			|input| {
+				if input.len() < 2 {
+					return None;
+				}
+				let mut trimmed_input = &input[2..];
+				ProviderHeader::decode(&mut trimmed_input).ok()
+			},
+		);
+		cfg_if::cfg_if! {
+			if #[cfg(feature = "runtime-benchmarks")] {
+				let provider_header = provider_header_result.unwrap_or_else(|_| ProviderHeader::new(<ProviderHeader as HeaderT>::Number::default(), <ProviderHeader as HeaderT>::Hash::default(), <ProviderHeader as HeaderT>::Hash::default(), <ProviderHeader as HeaderT>::Hash::default(), sp_runtime::Digest::default()));
+			} else {
+				let provider_header = provider_header_result.map_err(Error::ParaHeadMerkleProof)?;
+			}
+		}
```

Comment:
```suggestion
				let provider_header = {
			cfg_if::cfg_if! {
				if #[cfg(feature = "runtime-benchmarks")] {
					provider_header_result.unwrap_or_else(|_| ProviderHeader::new(<ProviderHeader as HeaderT>::Number::default(), <ProviderHeader as HeaderT>::Hash::default(), <ProviderHeader as HeaderT>::Hash::default(), <ProviderHeader as HeaderT>::Hash::default(), sp_runtime::Digest::default()))
				} else {
					 provider_header_result.map_err(Error::ParaHeadMerkleProof)?
				}
			}
		};
```

Reply:
I hope you can read it

Reply:
This is also not relevant because of the benchmarking PR.